### PR TITLE
Touch application form and choices when course changes

### PR DESCRIPTION
**Pull request by despo**
## Context

When we update course details in publish, and we sync those details through the TTAPI, those will be updated in Apply. However those changes don't make their way to the Vendor API responses as application choices are not touched. 

## Changes proposed in this pull request

When course details are updated, we touch application forms as well as choices in the TTAPI sync

## Link to Trello card

https://trello.com/c/g9PUXgZZ/4463-touch-application-form-and-choices-when-course-changes

## Things to check

- [X] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/models/course.rb [hunk 1]
```diff
@@ -162,4 +165,15 @@ def open!
...
   end
+
+private
+
+  def touch_application_choices_and_forms
+    return if saved_changes['start_date'].blank?
+
+    ActiveRecord::Base.transaction do
+      application_choices.touch_all
+      ApplicationForm.where(application_choices: application_choices).touch_all
```
Comment (steventux) [inline app/models/course.rb:176]: Doesn't this https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/models/application_choice.rb#L7 touch the form when the choices are touched?

**Review by steventux — COMMENTED** [review]
A question about the existing mechanism for touching `ApplicationForm`s via the choices.

File: app/models/course.rb (hunk 1 above)
Comment (despo) [reply app/models/course.rb:176]: No because `touch_all` doesn't trigger callbacks.

File: app/models/course.rb [hunk 2]
```diff
@@ -162,4 +165,15 @@ def open!
...
     )
   end
+
+private
+
+  def touch_application_choices_and_forms
+    return if saved_changes['start_date'].blank?
+
+    ActiveRecord::Base.transaction do
+      application_choices.touch_all
```
Comment (cwrw) [inline app/models/course.rb:175]: could we also make sure those are only in the current cycle?

File: app/models/course.rb [hunk 3]
```diff
@@ -18,6 +18,9 @@ class Course < ApplicationRecord
   scope :in_cycle, ->(year) { where(recruitment_cycle_year: year) }
 
   scope :with_course_options, -> { left_outer_joins(:course_options).where.not(course_options: { id: nil }) }
+
+  after_update :touch_application_choices_and_forms, if: :in_current_recruitment_cycle?
```
Comment (despo) [inline app/models/course.rb:22]: @cwrw ^

**Review by cwrw — APPROVED** [review]
looks good, could you add a test for when its not in the same cycle and also when start date isn't updated just to be sure?

**Review by steventux — APPROVED** [review]

**Pull request merged**